### PR TITLE
Add a per-light volumetric fog energy property

### LINKED ISSUE #1374: Add a volumetric cull mask to lights for volumetric fog (and implement an "additive fog" option for volumetric lighting)
*Note: This feature was discussed with reduz. It's probably good to implement if someone finds a way to do it.*

**Describe the project you are working on:**

The Godot editor :slightly_smiling_face: 

**Describe the problem or limitation you are having in your project:**

With the [newly-added volumetric fog effect](https://github.com/godotengine/godot/pull/41213), it's not possible to prevent individual lights from contributing to the volumetric fog's color.

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**

A volumetric cull mask property could be added to prevent individual lights from contributing to the volumetric fog's color.

Alternatively, this can be done by adding a Volumetric Fog Energy property to Light3D. (It's probably better as it gives more control to the user, rather than an all-or-nothing solution.)

In addition, there should probably be a way to disable fog completely in areas not lit by anything (instead of having a black fog). This way, it can be used purely as a volumetric lighting solution, not adding any fog unless desired by the artist.

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**

There could be a Volumetric Cull Mask property mimicking Light's Cull Mask property. It would only affect volumetric fog. Direct and indirect lighting would be left unaffected by this mask.

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**

No, as this is core renderer functionality.

**Is there a reason why this should be core and not an add-on in the asset library?:**

This is core renderer functionality.

### PR DESCRIPTION
**Pull request by Calinou**
Per-light energy gives more control to the user on the final result of volumetric fog. Specific lights can be fully excluded from volumetric fog by setting their volumetric fog energy to 0, which improves performance slightly. This can also be used to prevent short-lived dynamic effects from poorly interacting with volumetric fog, as it's updated over several frames by default unless temporal reprojection is disabled.

Volumetric fog shadows now obey Light3D's Shadow Opacity property as well.

The shadow fog fade property was removed as it had little visible impact on the final scene's rendering. Its default value (`0.1 = 1/0.1 => 10.0`) is now hardcoded within the shader, but is kept as a constant for tinkering in the future.

You can now use volumetric fog as an approximation of a volumetric *lighting* solution by following these steps:

- Enable volumetric fog in WorldEnvironment.
- Set the fog density to a very low but **non-zero** value like `0.002`. (`0.001` will not work as it's the threshold for disabling volumetric fog in the shader.)
- Set Volumetric Fog Energy on your lights to much higher values compared to the default (typically between `100` and `10000`).
- You can still use FogVolume nodes at the same time, as they have their own density defined.

This closes https://github.com/godotengine/godot-proposals/issues/1374.

**Testing project:** [test_volumetric_fog_per_light.zip](https://github.com/godotengine/godot/files/9176867/test_volumetric_fog_per_light.zip)

## Preview

https://user-images.githubusercontent.com/180032/180669307-64a6cf77-6c71-4b7c-ae79-eec7ca52bee7.mp4

### CODE REVIEW

File: scene/3d/light_3d.cpp [hunk 1]
```diff
@@ -329,7 +329,6 @@ void Light3D::_bind_methods() {
 	BIND_ENUM_CONSTANT(PARAM_SHADOW_BIAS);
 	BIND_ENUM_CONSTANT(PARAM_SHADOW_PANCAKE_SIZE);
 	BIND_ENUM_CONSTANT(PARAM_SHADOW_BLUR);
-	BIND_ENUM_CONSTANT(PARAM_SHADOW_VOLUMETRIC_FOG_FADE);
```
Comment (clayjohn) [inline scene/3d/light_3d.cpp:332]: You will need to bind the new enum and update docs

**Review by clayjohn — CHANGES_REQUESTED** [review]

File: scene/3d/light_3d.cpp (hunk 1 above)
Comment (Calinou) [reply scene/3d/light_3d.cpp:332]: Done :slightly_smiling_face:

**Comment by akien-mga** [comment]
Needs a rebase. @reduz says that if Clay is OK this should be fine to merge.

**Review by clayjohn — APPROVED** [review]
Looks good to me! Should be good to merge once conflicts are resolved

**Comment by Calinou** [comment]
Rebased and tested again, it works as expected.

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**